src/providers/eth_provider/provider.rs

```diff
@@ -189,35 +199,79 @@ where
+        };
+        let data = request.input.into_input().unwrap_or_default();
+        let tx = TxLegacy {
+            chain_id: Some(*ETH_CHAIN_ID),
+            nonce: request.nonce.unwrap_or_default(),
+            gas_price,
+            gas_limit: request.gas.unwrap_or_default(),
+            to: request.to.unwrap_or_default(),
+            value: request.value.unwrap_or_default(),
+            input: data,
+        };
+        let ser_tx = CairoArrayLegacy(tx.encoded_for_signing().into_iter().map(Felt::from).collect::<Vec<_>>());
+        let ser_tx_len = Felt::from(ser_tx.len());
+
+        // Init a connected account to the first relayer in the set
+        let block_id = self.to_starknet_block_id(block_id).await?;
+        let wallet = LocalWallet::from_signing_key(SigningKey::from_random());
+        let mut connected_account = SingleOwnerAccount::new(
+            self.starknet_provider_inner(),
+            wallet,
+            RELAYERS.first().copied().expect("always at least one relayer"),
+            self.chain_id.into(),
+            ExecutionEncoding::New,
+        );
+        connected_account.set_block_id(block_id);
+
+        let kakarot_contract = KakarotCore::new(*KAKAROT_ADDRESS, connected_account);
+        let span = tracing::span!(tracing::Level::INFO, "sn::eth_send_raw_unsigned_tx");
+
+        // Simulate the execution of an unsigned transaction
+        let res = kakarot_contract
+            .eth_send_raw_unsigned_tx(&ser_tx_len, &ser_tx)
+            .simulate(true, true)
             .instrument(span)
             .await
-            .map_err(ExecutionError::from)?;
+            .map_err(|_| ExecutionError::CairoVm(CairoError::VmOutOfResources))?;
+
+        let res = match res.transaction_trace {
+            // Return the result if the transaction passes
+            TransactionTrace::Invoke(InvokeTransactionTrace {
+                execute_invocation: ExecuteInvocation::Success(FunctionInvocation { result, .. }),
+                ..
+            }) => result,
+            // Return a cairo vm out of steps error in case we reverted with "out of steps"
+            TransactionTrace::Invoke(InvokeTransactionTrace {
+                execute_invocation: ExecuteInvocation::Reverted(RevertedInvocation { revert_reason }),
+                ..
+            }) if revert_reason.contains("RunResources has no remaining steps") => {
+                return Err(ExecutionError::CairoVm(CairoError::VmOutOfResources).into());
+            }
+            // Return the error as normal otherwise.
+            TransactionTrace::Invoke(InvokeTransactionTrace {
+                execute_invocation: ExecuteInvocation::Reverted(RevertedInvocation { revert_reason }),
+                ..
+            }) => {
+                return Err(TransactionError::Call(revert_reason.into()).into());
+            }
+            _ => unreachable!(),
```

Comment:
are we sure we want to add such code in a server?

Reply:
unreachable isn't an error, it translates something that can never happen in your code. In this case, we will never match on this because the original simulation we make is an invoke, and we already match on all possibilities given the fact that we know we are simulating an `Invoke` and nothing else

Reply:
but if in another version, they add a new enum, we'll have a problem